src/Common/RingBuffer.h

```diff
@@ -0,0 +1,154 @@
+#pragma once
+
+#include <vector>
+#include <iostream>
+
+namespace DB
+{
+
+/**
+ * A ring buffer of fixed size.
+ * With an ability to expand / narrow.
+ * When narrowing only first N elements remain.
+ */
+
+template <class T>
+class RingBuffer
```

Comment:
A couple of ideas:
* It is possible to push and pop from both ends of the buffer. Maybe rename methods to `push_back()` and `pop_front()`, so that we could add `push_front()` and `pop_back()` later if needed.
* Maybe it would be better to avoid `try` methods, cause client actually can check if buffer `empty` or `full` beforehand. Or maybe add ordinary `push_` and `pop_`, which work using `try` and throw (or assert) if false.
* For push, it is better to use universal `&&` references and `std::forward` stuff.
* Maybe it is better not to return value from `pop`, but allow access `front()` and `back()`.
* Now, `T` should be default constructible. It is probably ok, but may be improved.

Reply:
I thought about all these ideas... But wrote a prototype, cause was not sure if it is really needed. I can make it more STL-like, but a bit later.

Reply:
Better to delete this and use `circular_buffer` from boost

Reply:
Deleted.